control_plane/attachment_service/src/tenant_state.rs

```diff
@@ -370,13 +370,44 @@ impl TenantState {
         // All remaining observed locations generate secondary intents.  This includes None
         // observations, as these may well have some local content on disk that is usable (this
         // is an edge case that might occur if we restarted during a migration or other change)
+        //
+        // We may leave intent.attached empty if we didn't find any attached locations: [`Self::schedule`]
+        // will take care of promoting one of these secondaries to be attached.
         self.observed.locations.keys().for_each(|node_id| {
             if Some(*node_id) != self.intent.attached {
                 self.intent.secondary.push(*node_id);
             }
         });
     }
 
+    /// Part of [`Self::schedule`] that is used to choose exactly one node to act as the
+    /// attached pageserver for a shard.
+    ///
+    /// Returns whether we modified it, and the NodeId selected.
+    fn schedule_attached(
+        &mut self,
+        scheduler: &mut Scheduler,
+    ) -> Result<(bool, NodeId), ScheduleError> {
+        // No work to do if we already have an attached tenant
+        if let Some(node_id) = self.intent.attached {
+            return Ok((false, node_id));
+        }
+
+        if let Some(promote_secondary) = scheduler.node_preferred(&self.intent.secondary) {
+            // Promote a secondary
+            tracing::debug!("Promoted secondary {} to attached", promote_secondary);
+            self.intent.secondary.retain(|n| n != &promote_secondary);
+            self.intent.attached = Some(promote_secondary);
+            Ok((true, promote_secondary))
+        } else {
+            // Pick a fresh node: either we had no secondaries or none were schedulable
+            let node_id = scheduler.schedule_shard(&self.intent.secondary)?;
+            tracing::debug!("Selected {} as attached", node_id);
+            self.intent.set_attached(scheduler, Some(node_id));
```

Comment:
nit: since there's no node already attached, `set_attached` is just an assignment. Can you make the if-else branches set the attachment consistently (assign in both or call `set_attached` in both)?

Reply:
This is a good observation: the reason I was directly setting these was because set_attached increments refcounts in the scheduler, but this secondary node was already referenced.  That was naughty though: IntentState is meant to wrap all the refcount logic (and in future this transition _will_ update a refcount, because we'll track attached + secondary counts separately).  I've added a commit that refactors it into an `IntentState::promote_attached` function and added a couple of TODOs to note where that future refcounting will be needed.